system-contracts/contracts/EvmInterpreter.yul

```diff
@@ -0,0 +1,1187 @@
+
+            function MEM_OFFSET_INNER() -> offset {
+                offset := add(MEM_OFFSET(), 32)
+            }
+
+            function MAX_POSSIBLE_MEM() -> max {
+                max := 0x100000 // 1MB
+            }
+
+            function MAX_MEMORY_FRAME() -> max {
+                max := add(MEM_OFFSET_INNER(), MAX_POSSIBLE_MEM())
+            }
+    
+            // It is the responsibility of the caller to ensure that ip >= BYTECODE_OFFSET + 32
+            function readIP(ip) -> opcode {
+                // TODO: Why not do this at the beginning once instead of every time?
+                let bytecodeLen := mload(BYTECODE_OFFSET())
+
+                let maxAcceptablePos := add(add(BYTECODE_OFFSET(), bytecodeLen), 31)
+                if gt(ip, maxAcceptablePos) {
+                    revert(0, 0)
+                }
+
+                opcode := and(mload(sub(ip, 31)), 0xff)
+            }
+
+            function readBytes(start, length) -> value {
+                let max := add(start, length)
+                for {} lt(start, max) { start := add(start, 1) } {
+                    let next_byte := readIP(start)
+
+                    value := or(shl(8, value), next_byte)
+                }
+            }
+
+            function dupStackItem(sp, evmGas, position) -> newSp, evmGasLeft {
+                evmGasLeft := chargeGas(evmGas, 3)
+                let tempSp := sub(sp, mul(0x20, sub(position, 1)))
+
+                if or(gt(tempSp, BYTECODE_OFFSET()), eq(tempSp, BYTECODE_OFFSET())) {
+                    revert(0, 0)
+                }
+                
+                if lt(tempSp, STACK_OFFSET()) {
+                    revert(0, 0)
+                }
+
+                let dup := mload(tempSp)                    
+
+                newSp := add(sp, 0x20)
+                mstore(newSp, dup)
+            }
+
+            function swapStackItem(sp, evmGas, position) ->  evmGasLeft {
+                evmGasLeft := chargeGas(evmGas, 3)
+                let tempSp := sub(sp, mul(0x20, position))
+
+                if or(gt(tempSp, BYTECODE_OFFSET()), eq(tempSp, BYTECODE_OFFSET())) {
+                    revert(0, 0)
+                }
+                
+                if lt(tempSp, STACK_OFFSET()) {
+                    revert(0, 0)
+                }
+
+                
+                let s2 := mload(sp)
+                let s1 := mload(tempSp)                    
+
+                mstore(sp, s1)
+                mstore(tempSp, s2)
+            }
+
+            function popStackItem(sp) -> a, newSp {
+                // We can not return any error here, because it would break compatibility
+                if lt(sp, STACK_OFFSET()) {
+                    revert(0, 0)
+                }
+
+                a := mload(sp)
+                newSp := sub(sp, 0x20)
+            }
+
+            function pushStackItem(sp, item) -> newSp {
+                if or(gt(sp, BYTECODE_OFFSET()), eq(sp, BYTECODE_OFFSET())) {
```

Comment:
Do you have an off-by-one here? I think you might be able to write to `BYTECODE_OFFSET()`.